src/box/index.cc

```diff
@@ -573,6 +573,109 @@ iterator_position(struct iterator *it, const char **pos, uint32_t *size)
 	return it->position(it, pos, size);
 }
 
+/** Keys for iterator position map. All the keys must be uint. */
+enum {
+	ITERATOR_POSITION_KEY = 0,
+};
+
+uint32_t
+iterator_position_pack_size(const char *pos, const char *pos_end,
+			    uint32_t part_count)
+{
+	assert(pos != NULL);
+	assert(part_count != 0);
+	uint32_t total = pos_end - pos;
+	total += mp_sizeof_array(part_count);
+	total += mp_sizeof_uint(ITERATOR_POSITION_KEY);
+	total += mp_sizeof_map(1);
+	total += mp_sizeof_binl(total);
+	return total;
+}
+
+void
+iterator_position_pack(const char *pos, const char *pos_end,
+		       uint32_t part_count, char *packed_pos,
+		       const char *packed_pos_end)
+{
+	(void)packed_pos_end;
+	assert(pos != NULL);
+	assert(pos_end != NULL);
+	assert(packed_pos != NULL);
+	assert(packed_pos_end != NULL);
+	assert(packed_pos < packed_pos_end);
+	assert(packed_pos_end - packed_pos >=
+	       iterator_position_pack_size(pos, pos_end, part_count));
+
+	uint32_t map_len = pos_end - pos;
+	map_len += mp_sizeof_array(part_count);
+	map_len += mp_sizeof_uint(ITERATOR_POSITION_KEY);
+	map_len += mp_sizeof_map(1);
+	packed_pos = mp_encode_binl(packed_pos, map_len);
+	packed_pos = mp_encode_map(packed_pos, 1);
+	packed_pos = mp_encode_uint(packed_pos, ITERATOR_POSITION_KEY);
+	packed_pos = mp_encode_array(packed_pos, part_count);
+	memcpy(packed_pos, pos, pos_end - pos);
+}
+
+int
+iterator_position_unpack(const char *packed_pos, const char *packed_pos_end,
+			 const char **pos, const char **pos_end,
+			 uint32_t *part_count)
+{
+	assert(packed_pos != NULL);
+	assert(packed_pos_end != NULL);
+	assert(pos != NULL);
+	assert(pos_end != NULL);
+	assert(part_count != NULL);
+
+	*pos = NULL;
+	*pos_end = NULL;
+	*part_count = 0;
+	uint32_t map_len = 0;
+	uint32_t bin_len = 0;
+	const char *array = NULL;
+	const char *packed_pos_check = NULL;
+	if (packed_pos + 1 >= packed_pos_end ||
+	    mp_typeof(*packed_pos) != MP_BIN ||
+	    mp_check_binl(packed_pos, packed_pos_end) > 0)
+		goto fail;
+	bin_len = mp_decode_binl(&packed_pos);
+	if (packed_pos + bin_len != packed_pos_end)
+		goto fail;
+	packed_pos_check = packed_pos;
+	if (mp_check(&packed_pos_check, packed_pos_end) != 0)
```

Comment:
Nit: Let's call `mp_check` before `mp_decode_binl`. Then you wouldn't need to use `mp_check_binl`.

Reply:
Data that binary contains will not be checked then.

Reply:
You'd still need the `mp_typeof(*packed_pos) != MP_BIN` check, but you wouldn't need the `packed_pos +1 >= packed_pos_end` and `mp_check_binl(packed_pos, packed_pos_end) != 0` checks. The code would be simpler and more consistent with the code below, where you don't use `mp_check_array`.

Reply:
Done

Reply:
Now, there's two `mp_check`... Also, you don't need to check that `packed_pos + bin_len == packed_pos_end` - instead you can just check that `packed_pos_check == packed_pos_end` after calling `mp_check`, which would be more straightforward.

Reply:
Right.